url4df616fce9_s3e0n1g1_prototype3_tgz-pJ8-Trading_Engine_MainJ8/src/Selecting_Algothrim/lecMSMomentum.java

```diff
@@ -0,0 +1,384 @@
+			arrayIsFull = true;
+		}
+		// this makes counting equals to zero, so it works like a round array
+		if ( counting == SIZE_OF_ARRAY ) { // to see if counting reachs 9
+			counting = 0;
+		}
+		// insert the trade inside the tenArray
+		tenArray[counting] = trade;
+		counting++;
+		// take Average only when arrayIsFull
+		if ( arrayIsFull){
+			lastAverage = average;
+			average = 0;
+			for ( int i = 0 ; i < SIZE_OF_ARRAY ; i++ ){
+				//System.out.println(i + " value: " + tenArray[i]);
+				average += tenArray[i]; 
+			}
+			average = average/10;
+			averageChange = (average - lastAverage)/ lastAverage;
+		}
+	}
+	public LinkedList<signalObject> generateSignalList( LinkedList<orderObject> buyOrders, LinkedList<orderObject> sellOrders){
+		
+		int countingBuyList = 0;
+		int countingSellList = 0;
+		double tempPrice = 0;
+		int tempAmount;
+		LinkedList<signalObject> result = new LinkedList<signalObject>();
+		boolean requestHappened = false;
+		// share quantityLefit will be initialed here
+		this.shareQuantityLeft = this.shareQuantityToTrade;
+
+		
+		if ( arrayIsFull == true){  // fill the list if only there arrayIsfull
+			if ( mode == BUY_MODE){ // if we have to buy
+				if (this.averageChange >  epsilon ){
+					// we put buy sell here
+					requestHappened = true;// for later checking to see if we need to swap
+					
+					while ( this.shareQuantityLeft > 0){
+					
+						if (buyOrders.get(0).getPrice() >= sellOrders.get(countingSellList).getPrice() ){
+							tempPrice = buyOrders.get(0).getPrice() + this.epsilon; // what is wrong in here
+						}else{// means buy has a buy has a lower price than sell, so we just need to give a price equal to sell price
+							// we don't need to beat the buy price
+							tempPrice = sellOrders.get(countingSellList).getPrice();
+						}// 
+						// we have price so far
+						// now we have to choose for the amount
+						if ( sellOrders.get(countingSellList).getQauntity() < this.shareQuantityLeft ){// if amount of orderObject is less than what we need 
+							// we need to use that and other shares
+							tempAmount = sellOrders.get(countingSellList).getQauntity();
+							result.add( new signalObject(tempAmount, tempPrice, "buy"  ));// we make a buy signal
+							this.shareQuantityLeft -= sellOrders.get(countingSellList).getQauntity(); // we decrease amount that we have to buy
+							countingSellList++; // we have to check for the  other one
+						}else{ // means that the amount we want to buy is less than what we other person wants to sell
+							tempAmount = this.shareQuantityLeft;
+							result.add(new signalObject(tempAmount, tempPrice, "buy"));
+							this.shareQuantityLeft = 0;
+							countingSellList++;
+						}
+					} // is is for while loop, means we do it till we put buy signal order into list till we buy amount we want
+				}// there is no else,  if average wasn't high enough we won't trade
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 73:100, 104:132, 292:318, 324:354.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.